test/index.test.ts

```diff
@@ -771,6 +771,182 @@ describe('Unit | Utility | changeset', () => {
     expect(dummyModel.name.short).toBe('foo');
   });
 
+  describe('arrays within nested objects', () => {
+    describe('#set', () => {
+      let initialData: { contact: { emails?: string[] } } = { contact: { emails: [] } };
+
+      beforeEach(() => {
+        initialData = { contact: { emails: ['bob@email.com'] } };
+      });
+
+      it('nested objects cannot create arrays when we have no hints', () => {
+        initialData.contact = {};
+
+        const changeset = Changeset(initialData);
+        expect(changeset.get('contact.emails')).toEqual(undefined);
+
+        changeset.set('contact.emails.0', 'fred@email.com');
+        expect(changeset.get('contact.emails.0')).toEqual('fred@email.com');
+        expect(changeset.get('contact.emails')).toEqual({ '0': 'fred@email.com' });
+      });
+
+      it('works with validations', () => {
+        const changeset = Changeset(
+          initialData,
+          lookupValidator({
+            contact: {
+              emails: [
+                (_k: string, value: any) => {
+                  if (value.includes('fred')) {
+                    return 'Fred is banned';
+                  }
+                }
+              ]
+            }
+          })
+        );
+
+        expect(changeset.isValid).toEqual(true);
+
+        changeset.set('contact.emails.0', 'fred@email.com');
+
+        expect(changeset.isValid).toEqual(false);
+        expect(changeset.errors).toEqual([
+          { key: 'contact.emails.0', validation: 'Fred is banned', value: 'fred@email.com' }
+        ]);
+      });
+
+      it('can be rolled back', () => {
+        const changeset = Changeset(initialData);
+
+        changeset.set('contact.emails.0', 'fred@email.com');
+
+        expect(changeset.get('contact.emails.0')).toEqual('fred@email.com');
+        expect(changeset.changes).toEqual([{ key: 'contact.emails.0', value: 'fred@email.com' }]);
+        expect(changeset.get('contact.emails')).toEqual(['fred@email.com']);
```

Comment:
without the weird jest hacks, this would require .unwrap() to be equallable with an array.
I think the main issue comes from the original object the proxy wraps being an object -- I tried to find a way to change that with Proxy, but it seems the ability to change how an object is iterated over or perceived is deliberately not part of the Proxy spec -- which kinda makes sense, cause it'd probs be super confusing

Reply:
it would! _most excellent_

Reply:
and yeah, I don't know of any situations where in-app, I'd want to access a non-leaf